dvc/dependency/param.py

```diff
@@ -37,51 +39,82 @@ class ParamsDependency(Dependency):
-                info = {self.PARAM_PARAMS: params}
-
-        super().__init__(
-            stage,
-            path
-            or os.path.join(stage.repo.root_dir, self.DEFAULT_PARAMS_FILE),
-            info=info,
-            repo=repo,
+        self.params = list(params) if params else []
+        info = (
+            {self.PARAM_PARAMS: params} if isinstance(params, dict) else None
         )
+        repo = repo or stage.repo
+        path = path or os.path.join(repo.root_dir, self.DEFAULT_PARAMS_FILE)
+        super().__init__(stage, path, info=info, repo=repo)
 
     def dumpd(self):
         ret = super().dumpd()
         if not self.hash_info:
-            ret[self.PARAM_PARAMS] = self.params
+            ret[self.PARAM_PARAMS] = self.params or {}
         return ret
 
     def fill_values(self, values=None):
         """Load params values dynamically."""
-        if not values:
+        if values is None:
             return
+
         info = {}
+        if not self.params:
+            info.update(values)
         for param in self.params:
             if param in values:
                 info[param] = values[param]
         self.hash_info = HashInfo(self.PARAM_PARAMS, info)
 
+    def read_params(
+        self, flatten: bool = True, **kwargs: typing.Any
+    ) -> Dict[str, typing.Any]:
+        try:
+            config = self.read_file()
+        except MissingParamsFile:
+            config = {}
+
+        if not self.params:
+            return config
+
+        ret = {}
+        if flatten:
+            for param in self.params:
+                try:
+                    ret[param] = dpath.util.get(config, param, separator=".")
+                except KeyError:
+                    continue
+            return ret
+
+        from dpath.util import merge
+
+        for param in self.params:
+            merge(
+                ret,
+                dpath.util.search(config, param, separator="."),
+                separator=".",
+            )
+        return ret
+
     def workspace_status(self):
-        status = super().workspace_status()
+        if not self.exists:
+            return {str(self): "deleted"}
+        if self.hash_info.value is None:
+            return {str(self): "new"}
 
-        if status.get(str(self)) == "deleted":
-            return status
+        from funcy import ldistinct
 
         status = defaultdict(dict)
         info = self.hash_info.value if self.hash_info else {}
         actual = self.read_params()
-        for param in self.params:
-            if param not in actual.keys():
+
+        # NOTE: we want to preserve the order of params as specified in the
+        # status. In case of tracking the whole file, the order is top-level
+        # keys in the file and then the keys in the `info` from `dvc.lock`
+        # (which are alphabetically sorted).
+        params = self.params or ldistinct([*actual.keys(), *info.keys()])
+        for param in params:
```

Comment:
Rather than violate this principle, I'd prefer to just keep `dvc status` showing changes at the file level. Again, I think it's similar to how param keys are handled today. If I have a stage with param `foo` and modify the value of `foo.bar`, `dvc status` shows `foo` is modified, not `foo.bar`. Similarly, I think for tracking `params.yaml:`, `dvc status` should show `params.yaml` is modified, not any individual parameter key.

Reply:
I did consider only showing `modified` on the file-level itself, but I think granular tracking of params to be an important feature that makes it distinct from other dependencies/outputs.

I did this only in `status` because `status` is not a static operation, but a very active operation that inspects the user's workspace. So it already has a context of what things are in the user's hyperparameters, so it seemed fine to also look into the lockfile (that helps us understand what was actually tracked previously so that we can provide `new`/`deleted` statuses which we kind of already did). 
So, I consider violating that to be a minor issue here in the `status`.


> dvc status should show params.yaml is modified, not any individual parameter key.

But again, I don't have a strong opinion here. Let me know if you feel that strongly here, should be a straightforward change.

Reply:
> that helps us understand what was actually tracked previously so that we can provide `new`/`deleted` statuses which we kind of already did

Yeah, now that I think about it more, I'm not that clear on how this is different from what's already happening, which seems to be comparing parameters in `dvc.yaml` to `dvc.lock`.

> But again, I don't have a strong opinion here. Let me know if you feel that strongly here, should be a straightforward change.

I would opt for whatever is the simplest and closest to the current implementation. We are about to do much larger changes in `status` soon, including separating data and stage status. We can revisit then when we have a clearer picture of what we want to do with this info.

Reply:
> Yeah, now that I think about it more, I'm not that clear on how this is different from what's already happening, which seems to be comparing parameters in `dvc.yaml` to `dvc.lock`.

yes we already did that, but here in this case we don't know what parameters we are supposed to be comparing without actually looking into hyperparameters file and `dvc.lock` combined. As I said, it's a very minor issue and not very relevant in the `status` operation.

Reply:
> I would opt for whatever is the simplest and closest to the current implementation. We are about to do much larger changes in `status` soon, including separating data and stage status. We can revisit then when we have a clearer picture of what we want to do with this info.

As we have this already implemented, we can just go with what we have. Like you said, we can revisit this without any issues in the future.